bliss/models/galsim_decoder.py

```diff
@@ -0,0 +1,291 @@
+            components.append(bulge)
+        galaxy = galsim.Add(components)
+        gal_conv = galsim.Convolution(galaxy, psf)
+        offset = offset if offset is None else offset.numpy()
+        image = gal_conv.drawImage(
+            nx=slen, ny=slen, method="auto", scale=self.pixel_scale, offset=offset
+        )
+        return torch.from_numpy(image.array).reshape(1, slen, slen)
+
+
+class UniformGalsimGalaxiesPrior:
+    def __init__(
+        self,
+        single_galaxy_prior: SingleGalsimGalaxyPrior,
+        max_n_sources: int,
+        max_shift: float,
+    ):
+        self.single_galaxy_prior = single_galaxy_prior
+        self.max_shift = max_shift  # between 0 and 0.5, from center.
+        self.max_n_sources = max_n_sources
+        self.dim_latents = self.single_galaxy_prior.dim_latents
+        assert 0 <= self.max_shift <= 0.5
+
+    def sample(self) -> Dict[str, Tensor]:
+        """Returns a single batch of source parameters."""
+        n_sources = _sample_n_sources(self.max_n_sources)
+
+        params = torch.zeros(self.max_n_sources, self.dim_latents)
+        params[:n_sources, :] = self.single_galaxy_prior.sample(n_sources)
+
+        locs = torch.zeros(self.max_n_sources, 2)
+        locs[:n_sources, 0] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+        locs[:n_sources, 1] = _uniform(-self.max_shift, self.max_shift, n_sources) + 0.5
+
+        # for now, galaxies only
+        galaxy_bools = torch.zeros(self.max_n_sources, 1)
+        galaxy_bools[:n_sources, :] = 1
+        star_bools = torch.zeros(self.max_n_sources, 1)
+
+        return {
+            "n_sources": torch.tensor(n_sources),
+            "galaxy_params": params,
+            "locs": locs,
+            "galaxy_bools": galaxy_bools,
+            "star_bools": star_bools,
+        }
+
+
+class UniformBrightestCenterGalsimGalaxy(UniformGalsimGalaxiesPrior):
+    def sample(self) -> Dict[str, Tensor]:
+        """Returns a single batch of source parameters where brightest galaxy is centered."""
+        sample = super().sample()
+        n_sources = sample.pop("n_sources")
+        flux = sample["galaxy_params"][:, 0].reshape(-1)
+        idx_order = torch.argsort(-flux)
+        reordered_sample = {k: v[idx_order] for k, v in sample.items()}
+        reordered_sample["locs"][0, :] = 0.5  # centered
+        return {"n_sources": n_sources, **reordered_sample}
+
+
+class GalsimGalaxiesDecoder:
+    def __init__(
+        self, single_galaxy_decoder: SingleGalsimGalaxyDecoder, slen: int, bp: int
+    ) -> None:
+        self.single_decoder = single_galaxy_decoder
+        self.slen = slen
+        self.bp = bp
+        assert self.slen + 2 * self.bp >= self.single_decoder.slen
+
+    def __call__(self, full_cat: FullCatalog):
+        return self.render_galaxies(full_cat, self.single_decoder.psf)
+
+    def render_galaxies(self, full_cat: FullCatalog, psf: galsim.GSObject):
```

Comment:
Call this "render_catalog" instead? Might as well if you are extending to stars eventually